server: fix instance IDs for shared-process tenant servers to node IDs

stevendanna opened the pull request:
We have a number of callers that assume the node ID and instance ID
can be used interchangably. In shared-process mode, this has revealed
a number of issues that we need to investigate and resolve.

This PR, however, papers over those issues to an extent by allowing
the shared process tenant servers to inherit the node ID from the
server that is starting it. This then ensures that their SQL instance
ID is the same as their node ID.

I've updated some of the names in NodeIDContainer to make the intended
usage more clear.

Fixes #84602

Epic: CRDB-14537

Release note: None

stevendanna committed:
server: small initialization cleanup

The podDialer is already correctly initialized by this point so we can
just use it directly.

Release note: None

stevendanna committed:
server: fix instance shared-process tenant servers to node IDs

We have a number of callers that assume the node ID and instance ID
can be used interchangably. In shared-process mode, this has revealed
a number of issues that we need to investigate and resolve.

This PR, however, papers over those issues to an extent by allowing
the shared process tenant servers to inherit the node ID from the
server that is starting it.

I've updated some of the names in NodeIDContainer to make the intended
usage more clear.

Release note: None

stevendanna committed:
server: don't capture logtags in tenant start ctx

As far as I can see, the only log tag that this was propogating
"tenant-orchestration". Since the context here is captured in a few
long-running processes, that task is likely more confusing than
helpful in the long run.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96857)
<!-- Reviewable:end -->

stevendanna committed:
server: avoid leaking tenant stopper

We've seen at least one stopper leak during a stress run. I was not
able to reproduce it locally, but from the code it seems possible that
we are leaking it here if something like this happens.

1. serverController's scanner ticks and starts scanning for expected
   tenants.

2. The server's stopper stops.

3. The tenant stopper for some tenant, foo, is also stopped in
   response. This tenant is stopped and removed from the running tenant
   entries.

4. Back in the serverController, we see that tenant foo isn't
   started. We try to start it.

5. We make a new stopper, but can't run an async task, so we exit with
   an error, leaving the stopper.

Epic: none

Release note: None

stevendanna commented on pkg/ccl/serverccl/server_controller_test.go:
```diff
@@ -181,6 +182,42 @@ VALUES($1, $2, $3, $4, $5)`, id, secret, username, created, expires)
 	require.Equal(t, body.Sessions[0].ApplicationName, "hello system")
 }
 
+func TestServerControllerMultiNodeTenantStartup(t *testing.T) {
```
This test fails under stress on master any time we get a mismatched node ID vs sql instance ID.

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just have a clarification question below.

> Reviewed 1 of 1 files at r1, 8 of 8 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96857#-:-NNw1uTG0vsU3lL5wwOO:b7w57s9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @smg260, @srosenberg, and @stevendanna)

___
*[`pkg/server/server_sql.go` line 1421 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96857#-NNw-z0s5eVrDPydRUm8:-NNw-z0s5eVrDPydRUm9:b-55o3wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8d993d83ae76a656637c44e222ec684b5da2f18/pkg/server/server_sql.go#L1421)):*
> ```Go
> 	// instance ID for a while.
> 	log.Infof(ctx, "bound sqlinstance: %v", instance)
> 	if err := s.sqlIDContainer.SetSQLInstanceID(ctx, instance.InstanceID); err != nil {
> ```

Bit confused about why this line has to run if the above `Create{Node}Instance` lines have run. Don't we already have an ID set by this point?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96857)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @smg260, and @srosenberg)

___
*[`pkg/server/server_sql.go` line 1421 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96857#-NNw-z0s5eVrDPydRUm8:-NO8mlnI8PH1RJWkIbOJ:b-t797k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8d993d83ae76a656637c44e222ec684b5da2f18/pkg/server/server_sql.go#L1421)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Bit confused about why this line has to run if the above `Create{Node}Instance` lines have run. Don't we already have an ID set by this point?
</blockquote></details>

In the case of a standalone mt sql pod we won't have an ID set by this point.

In the case of a mixed sql pod we will have a node ID already.  We don't need to call this in that case. This change came from a branch in which I also hope to separate the node ID and instance ID. I've left it such that we are still calling it just as a safety check that we are always setting it to the same value.


<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=dhartunian
